File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/StringConverter.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresReader;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public abstract class StringConverter {

	public static void serializeURI(PostgresBuffer sw, String value) {
		if (value == null) return;
		sw.addToBuffer(value);
	}

	public static void serializeCompositeURI(PostgresBuffer sw, String value) {
		if (value == null) return;
		for (int i = 0; i < value.length(); i++) {
			char c = value.charAt(i);
			if (c == '\\' || c == '/') {
				sw.addToBuffer('\\');
			}
			sw.addToBuffer(c);
		}
	}

	public static void skip(PostgresReader reader, int context) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return;
		}
		if (cur != '"' && cur != '\\') {
			reader.initBuffer();
			reader.fillUntil(',', ')');
			reader.read();
		} else {
			cur = reader.read(context);
			while (cur != -1) {
				if (cur == '\\' || cur == '"') {
					cur = reader.read(context);
					if (cur == ',' || cur == ')') {
						return;
					}
					cur = reader.read(context);
				} else cur = reader.read();
			}
			throw new IOException("Unable to find end of string");
		}
	}

	public static String parse(PostgresReader reader, int context, boolean allowNulls) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return allowNulls ? null : "";
		}
		if (cur != '"' && cur != '\\') {
			reader.initBuffer((char) cur);
			reader.fillUntil(',', ')');
			reader.read();
			return reader.bufferToString();
		}
		return parseEscapedString(reader, context, ')');
	}

	static String parseEscapedString(PostgresReader reader, int context, char matchEnd) throws IOException {
		int cur = reader.read(context);
		reader.initBuffer();
		do {
			if (cur == '\\' || cur == '"') {
				cur = reader.read(context);
				if (cur == ',' || cur == matchEnd) {
					return reader.bufferToString();
				}
				for (int i = 0; i < context - 1; i++) {
					cur = reader.read();
				}
			}
			reader.addToBuffer((char) cur);
			reader.fillUntil('\\', '"');
			cur = reader.read();
		} while (cur != -1);
		throw new IOException("Unable to find end of string");
	}

	public static List<String> parseCollection(PostgresReader reader, int context, boolean allowNull) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean escaped = cur != '{';
		if (escaped) {
			reader.read(context);
		}
		cur = reader.peek();
		if (cur == '}') {
			if (escaped) {
				reader.read(context + 2);
			} else {
				reader.read(2);
			}
			return new ArrayList<>(0);
		}
        int innerContext = context << 1;
		List<String> list = new ArrayList<>();
		String emptyCol = allowNull ? null : "";
		do {
			cur = reader.read();
			if (cur == '"' || cur == '\\') {
				list.add(parseEscapedString(reader, innerContext, '}'));
				cur = reader.last();
			} else {
				reader.initBuffer((char) cur);
				reader.fillUntil(',', '}');
				cur = reader.read();
				if (reader.bufferMatches("NULL")) {
					list.add(emptyCol);
				} else {
					list.add(reader.bufferToString());
				}
			}
		} while (cur == ',');
		if (escaped) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static PostgresTuple toTuple(String value) {
		return ValueTuple.from(value);
	}

}

File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/StringConverter.scala
package net.revenj.database.postgres.converters

import java.io.IOException

import net.revenj.database.postgres.{PostgresBuffer, PostgresReader}

object StringConverter extends Converter[String] {
  override def serializeURI(sw: PostgresBuffer, value: String): Unit = {
    if (value != null) sw.addToBuffer(value)
  }

  def serializeCompositeURI(sw: PostgresBuffer, value: String): Unit = {
    if (value != null) {
      var i = 0
      while (i < value.length) {
        val c = value.charAt(i)
        if (c == '\\' || c == '/') {
          sw.addToBuffer('\\')
        }
        sw.addToBuffer(c)
        i += 1
      }
    }
  }

  val dbName = "varchar"
  def default() = ""

  def skip(reader: PostgresReader, context: Int): Unit = {
    var cur = reader.read()
    if (cur != ',' && cur != ')') {
      if (cur != '"' && cur != '\\') {
        reader.initBuffer()
        reader.fillUntil(',', ')')
        reader.read()
      } else {
        cur = reader.read(context)
        while (cur >= 0) {
          if (cur == '\\' || cur == '"') {
            cur = reader.read(context)
            if (cur == ',' || cur == ')') {
              cur = -2
            } else {
              cur = reader.read(context)
            }
          } else cur = reader.read()
        }
        if (cur == -1) {
          throw new IOException("Unable to find end of string")
        }
      }
    }
  }

  override def parseRaw(reader: PostgresReader, start: Int, context: Int): String = {
    if (start != '"' && start != '\\') {
      reader.initBuffer(start.toChar)
      reader.fillUntil(',', ')')
      reader.read()
      reader.bufferToString()
    } else {
      parseEscapedString(reader, context, ')')
    }
  }

  private[converters] def parseEscapedString(reader: PostgresReader, context: Int, matchEnd: Char): String = {
    var cur = reader.read(context)
    reader.initBuffer()
    do {
      if (cur == '\\' || cur == '"') {
        cur = reader.read(context)
        if (cur == ',' || cur == matchEnd) {
          return reader.bufferToString()
        }
        var i: Int = 0
        while (i < context - 1) {
          cur = reader.read()
          i += 1
        }
      }
      reader.addToBuffer(cur.toChar)
      reader.fillUntil('\\', '"')
      cur = reader.read()
    } while (cur != -1)
    throw new IOException("Unable to find end of string")
  }

  override def parseCollectionItem(reader: PostgresReader, context: Int): String = {
    val cur = reader.read()
    if (cur == '"' || cur == '\\') {
      parseEscapedString(reader, context, '}')
    } else {
      reader.initBuffer(cur.toChar)
      reader.fillUntil(',', '}')
      reader.read()
      if (reader.bufferMatches("NULL")) {
        ""
      } else {
        reader.bufferToString()
      }
    }
  }
  override def parseNullableCollectionItem(reader: PostgresReader, context: Int): Option[String] = {
    val cur = reader.read()
    if (cur == '"' || cur == '\\') {
      Some(parseEscapedString(reader, context, '}'))
    } else {
      reader.initBuffer(cur.toChar)
      reader.fillUntil(',', '}')
      reader.read()
      if (reader.bufferMatches("NULL")) {
        None
      } else {
        Some(reader.bufferToString())
      }
    }
  }

  override def toTuple(value: String): PostgresTuple = ValueTuple.from(value)
}
